source/common/router/router.cc

```diff
@@ -747,6 +747,16 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
   UpstreamRequestPtr upstream_request =
       std::make_unique<UpstreamRequest>(*this, std::move(generic_conn_pool), can_send_early_data,
                                         can_use_http3, false /*enable_half_close*/);
+  if (retry_state_ && retry_state_->shouldRetryOnConnectionFailure() && ensure_connection_retry_) {
+    upstream_request->addUpstreamCallbacks(*this);
+    wait_for_connect_ = true;
```

Comment:
don't we also need to set this at doRetry?
If we have a request with headers, get a connection, get a fast-fail, kick of a retry, this should be true again right?

Reply:
I just wonder if we could have an issue in case of per try timeout:
1. request 1 per try timeout with a connection
2. router create request 2
3. request 1 call encodeHeaders

we would have `waiting_for_upstream_connection_ = true` when it would actually already have a connection with an upstream

My intention here was to ensure connection retry mostly until we get at least 1 connection. I believe we should us the `retry_shadow_buffer_limit_` for the other use case. 

Let me know if that make sense.

Reply:
I still don't understand why we can't simply use the cluster limit - why do we need to add to it?

Reply:
We can't use the cluster limit has the router might received more data than the cluster limit.
The upstream request will pause read when it has pass the buffer limit. So we need to keep a margin